File: binding/include/charts/libqlegendmarker.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qlegendmarker.html
pub const qlegendmarker = struct {

    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QLegendMarker_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qlegendmarker.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Type(self: ?*anyopaque, ) i32 {
        return qtc.QLegendMarker_Type(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnType(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QLegendMarker_OnType(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseType(self: ?*anyopaque, ) i32 {
        return qtc.QLegendMarker_QBaseType(@ptrCast(self));
    }

    pub fn Label(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QLegendMarker_Label(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qlegendmarker.Label: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetLabel(self: ?*anyopaque, label: []const u8) void {
        const label_str = qtc.libqt_string{
    .len = label.len,
    .data = label.ptr,
};
qtc.QLegendMarker_SetLabel(@ptrCast(self), label_str);
    }

    pub fn LabelBrush(self: ?*anyopaque, ) QtC.QBrush {
        return qtc.QLegendMarker_LabelBrush(@ptrCast(self));
    }

    pub fn SetLabelBrush(self: ?*anyopaque, brush: ?*anyopaque) void {
        qtc.QLegendMarker_SetLabelBrush(@ptrCast(self), @ptrCast(brush));
    }

    pub fn Font(self: ?*anyopaque, ) QtC.QFont {
        return qtc.QLegendMarker_Font(@ptrCast(self));
    }

    pub fn SetFont(self: ?*anyopaque, font: ?*anyopaque) void {
        qtc.QLegendMarker_SetFont(@ptrCast(self), @ptrCast(font));
    }

    pub fn Pen(self: ?*anyopaque, ) QtC.QPen {
        return qtc.QLegendMarker_Pen(@ptrCast(self));
    }

    pub fn SetPen(self: ?*anyopaque, pen: ?*anyopaque) void {
        qtc.QLegendMarker_SetPen(@ptrCast(self), @ptrCast(pen));
    }

    pub fn Brush(self: ?*anyopaque, ) QtC.QBrush {
        return qtc.QLegendMarker_Brush(@ptrCast(self));
    }

    pub fn SetBrush(self: ?*anyopaque, brush: ?*anyopaque) void {
        qtc.QLegendMarker_SetBrush(@ptrCast(self), @ptrCast(brush));
    }

    pub fn IsVisible(self: ?*anyopaque, ) bool {
        return qtc.QLegendMarker_IsVisible(@ptrCast(self));
    }

    pub fn SetVisible(self: ?*anyopaque, visible: bool) void {
        qtc.QLegendMarker_SetVisible(@ptrCast(self), visible);
    }

    pub fn Shape(self: ?*anyopaque, ) i32 {
        return qtc.QLegendMarker_Shape(@ptrCast(self));
    }

    pub fn SetShape(self: ?*anyopaque, shape: i32) void {
        qtc.QLegendMarker_SetShape(@ptrCast(self), @intCast(shape));
    }

    pub fn Series(self: ?*anyopaque, ) QtC.QAbstractSeries {
        return qtc.QLegendMarker_Series(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnSeries(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QAbstractSeries) void {
        qtc.QLegendMarker_OnSeries(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSeries(self: ?*anyopaque, ) QtC.QAbstractSeries {
        return qtc.QLegendMarker_QBaseSeries(@ptrCast(self));
    }

    pub fn Clicked(self: ?*anyopaque, ) void {
        qtc.QLegendMarker_Clicked(@ptrCast(self));
    }

    pub fn OnClicked(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QLegendMarker_Connect_Clicked(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Hovered(self: ?*anyopaque, status: bool) void {
        qtc.QLegendMarker_Hovered(@ptrCast(self), status);
    }

    pub fn OnHovered(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) void) void {
        qtc.QLegendMarker_Connect_Hovered(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn LabelChanged(self: ?*anyopaque, ) void {
        qtc.QLegendMarker_LabelChanged(@ptrCast(self));
    }

    pub fn OnLabelChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QLegendMarker_Connect_LabelChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn LabelBrushChanged(self: ?*anyopaque, ) void {
        qtc.QLegendMarker_LabelBrushChanged(@ptrCast(self));
    }

    pub fn OnLabelBrushChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QLegendMarker_Connect_LabelBrushChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn FontChanged(self: ?*anyopaque, ) void {
        qtc.QLegendMarker_FontChanged(@ptrCast(self));
    }

    pub fn OnFontChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QLegendMarker_Connect_FontChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn PenChanged(self: ?*anyopaque, ) void {
        qtc.QLegendMarker_PenChanged(@ptrCast(self));
    }

    pub fn OnPenChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QLegendMarker_Connect_PenChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn BrushChanged(self: ?*anyopaque, ) void {
        qtc.QLegendMarker_BrushChanged(@ptrCast(self));
    }

    pub fn OnBrushChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QLegendMarker_Connect_BrushChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn VisibleChanged(self: ?*anyopaque, ) void {
        qtc.QLegendMarker_VisibleChanged(@ptrCast(self));
    }

    pub fn OnVisibleChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QLegendMarker_Connect_VisibleChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ShapeChanged(self: ?*anyopaque, ) void {
        qtc.QLegendMarker_ShapeChanged(@ptrCast(self));
    }

    pub fn OnShapeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QLegendMarker_Connect_ShapeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QLegendMarker_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qlegendmarker.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QLegendMarker_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qlegendmarker.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QLegendMarker_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/legendmarker.html#types
pub const enums = struct {
    pub const LegendMarkerType = enum {
        pub const LegendMarkerTypeArea: i32 = 0;
        pub const LegendMarkerTypeBar: i32 = 1;
        pub const LegendMarkerTypePie: i32 = 2;
        pub const LegendMarkerTypeXY: i32 = 3;
        pub const LegendMarkerTypeBoxPlot: i32 = 4;
        pub const LegendMarkerTypeCandlestick: i32 = 5;
    };

};

File: binding/src/charts/libqlegendmarker.cpp
#include <QAbstractSeries>
#include <QBrush>
#include <QFont>
#include <QLegend>
#include <QLegendMarker>
#include <QObject>
#include <QPen>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qlegendmarker.h>
#include "libqlegendmarker.h"
#include "libqlegendmarker.hxx"

libqt_string QLegendMarker_Tr(const char* s) {
	QString _ret = QLegendMarker::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QLegendMarker_Type(QLegendMarker* self) {
	return self->type();
}

libqt_string QLegendMarker_Label(const QLegendMarker* self) {
	QString _ret = self->label();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QLegendMarker_SetLabel(QLegendMarker* self, const libqt_string label) {
	self->setLabel(QString::fromUtf8(label.data, label.len));
}

QBrush* QLegendMarker_LabelBrush(const QLegendMarker* self) {
	return new QBrush(self->labelBrush());
}

void QLegendMarker_SetLabelBrush(QLegendMarker* self, const QBrush* brush) {
	self->setLabelBrush(*brush);
}

QFont* QLegendMarker_Font(const QLegendMarker* self) {
	return new QFont(self->font());
}

void QLegendMarker_SetFont(QLegendMarker* self, const QFont* font) {
	self->setFont(*font);
}

QPen* QLegendMarker_Pen(const QLegendMarker* self) {
	return new QPen(self->pen());
}

void QLegendMarker_SetPen(QLegendMarker* self, const QPen* pen) {
	self->setPen(*pen);
}

QBrush* QLegendMarker_Brush(const QLegendMarker* self) {
	return new QBrush(self->brush());
}

void QLegendMarker_SetBrush(QLegendMarker* self, const QBrush* brush) {
	self->setBrush(*brush);
}

bool QLegendMarker_IsVisible(const QLegendMarker* self) {
	return self->isVisible();
}

void QLegendMarker_SetVisible(QLegendMarker* self, bool visible) {
	self->setVisible(visible);
}

int QLegendMarker_Shape(const QLegendMarker* self) {
	return self->shape();
}

void QLegendMarker_SetShape(QLegendMarker* self, int shape) {
	self->setShape(static_cast<QLegend::MarkerShape>(shape));
}

QAbstractSeries* QLegendMarker_Series(QLegendMarker* self) {
	return self->series();
}

void QLegendMarker_Clicked(QLegendMarker* self) {
	self->clicked();
}

void QLegendMarker_Connect_Clicked(QLegendMarker* self, intptr_t slot) {
    void (*slotFunc)(QLegendMarker*) = reinterpret_cast<void (*)(QLegendMarker*)>(slot);
    QLegendMarker::connect(self, &QLegendMarker::clicked, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QLegendMarker_Hovered(QLegendMarker* self, bool status) {
	self->hovered(status);
}

void QLegendMarker_Connect_Hovered(QLegendMarker* self, intptr_t slot) {
    void (*slotFunc)(QLegendMarker*, bool) = reinterpret_cast<void (*)(QLegendMarker*, bool)>(slot);
    QLegendMarker::connect(self, &QLegendMarker::hovered, [self, slotFunc](bool status) {
	slotFunc(self, status);
    });
}

void QLegendMarker_LabelChanged(QLegendMarker* self) {
	self->labelChanged();
}

void QLegendMarker_Connect_LabelChanged(QLegendMarker* self, intptr_t slot) {
    void (*slotFunc)(QLegendMarker*) = reinterpret_cast<void (*)(QLegendMarker*)>(slot);
    QLegendMarker::connect(self, &QLegendMarker::labelChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QLegendMarker_LabelBrushChanged(QLegendMarker* self) {
	self->labelBrushChanged();
}

void QLegendMarker_Connect_LabelBrushChanged(QLegendMarker* self, intptr_t slot) {
    void (*slotFunc)(QLegendMarker*) = reinterpret_cast<void (*)(QLegendMarker*)>(slot);
    QLegendMarker::connect(self, &QLegendMarker::labelBrushChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QLegendMarker_FontChanged(QLegendMarker* self) {
	self->fontChanged();
}

void QLegendMarker_Connect_FontChanged(QLegendMarker* self, intptr_t slot) {
    void (*slotFunc)(QLegendMarker*) = reinterpret_cast<void (*)(QLegendMarker*)>(slot);
    QLegendMarker::connect(self, &QLegendMarker::fontChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QLegendMarker_PenChanged(QLegendMarker* self) {
	self->penChanged();
}

void QLegendMarker_Connect_PenChanged(QLegendMarker* self, intptr_t slot) {
    void (*slotFunc)(QLegendMarker*) = reinterpret_cast<void (*)(QLegendMarker*)>(slot);
    QLegendMarker::connect(self, &QLegendMarker::penChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QLegendMarker_BrushChanged(QLegendMarker* self) {
	self->brushChanged();
}

void QLegendMarker_Connect_BrushChanged(QLegendMarker* self, intptr_t slot) {
    void (*slotFunc)(QLegendMarker*) = reinterpret_cast<void (*)(QLegendMarker*)>(slot);
    QLegendMarker::connect(self, &QLegendMarker::brushChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QLegendMarker_VisibleChanged(QLegendMarker* self) {
	self->visibleChanged();
}

void QLegendMarker_Connect_VisibleChanged(QLegendMarker* self, intptr_t slot) {
    void (*slotFunc)(QLegendMarker*) = reinterpret_cast<void (*)(QLegendMarker*)>(slot);
    QLegendMarker::connect(self, &QLegendMarker::visibleChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QLegendMarker_ShapeChanged(QLegendMarker* self) {
	self->shapeChanged();
}

void QLegendMarker_Connect_ShapeChanged(QLegendMarker* self, intptr_t slot) {
    void (*slotFunc)(QLegendMarker*) = reinterpret_cast<void (*)(QLegendMarker*)>(slot);
    QLegendMarker::connect(self, &QLegendMarker::shapeChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

libqt_string QLegendMarker_Tr2(const char* s, const char* c) {
	QString _ret = QLegendMarker::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QLegendMarker_Tr3(const char* s, const char* c, int n) {
	QString _ret = QLegendMarker::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
int QLegendMarker_QBaseType(QLegendMarker* self) {
	auto* vqlegendmarker = dynamic_cast<VirtualQLegendMarker*>(self);
	if (vqlegendmarker && vqlegendmarker->isVirtualQLegendMarker) {
vqlegendmarker->setQLegendMarker_Type_IsBase(true);
	return vqlegendmarker->type();
}
}

// Auxiliary method to allow providing re-implementation
void QLegendMarker_OnType(QLegendMarker* self, intptr_t slot) {
	auto* vqlegendmarker = dynamic_cast<VirtualQLegendMarker*>(self);
	if (vqlegendmarker && vqlegendmarker->isVirtualQLegendMarker) {
vqlegendmarker->setQLegendMarker_Type_Callback(reinterpret_cast<VirtualQLegendMarker::QLegendMarker_Type_Callback>(slot));
}
}

// Base class handler implementation
QAbstractSeries* QLegendMarker_QBaseSeries(QLegendMarker* self) {
	auto* vqlegendmarker = dynamic_cast<VirtualQLegendMarker*>(self);
	if (vqlegendmarker && vqlegendmarker->isVirtualQLegendMarker) {
vqlegendmarker->setQLegendMarker_Series_IsBase(true);
	return vqlegendmarker->series();
}
}

// Auxiliary method to allow providing re-implementation
void QLegendMarker_OnSeries(QLegendMarker* self, intptr_t slot) {
	auto* vqlegendmarker = dynamic_cast<VirtualQLegendMarker*>(self);
	if (vqlegendmarker && vqlegendmarker->isVirtualQLegendMarker) {
vqlegendmarker->setQLegendMarker_Series_Callback(reinterpret_cast<VirtualQLegendMarker::QLegendMarker_Series_Callback>(slot));
}
}

void QLegendMarker_Delete(QLegendMarker* self) {
    delete self;
}

